duo-256.nix

```diff
@@ -0,0 +1,216 @@
+    CONFIG_CRYPTO_USER_API_HASH=y
+    CONFIG_CRYPTO_HMAC=y
+    CONFIG_DMIID=y
+    CONFIG_AUTOFS_FS=y
+    CONFIG_TMPFS_POSIX_ACL=y
+    CONFIG_TMPFS_XATTR=y
+    CONFIG_SECCOMP=y
+    CONFIG_BLK_DEV_INITRD=y
+    CONFIG_BINFMT_ELF=y
+    CONFIG_INOTIFY_USER=y
+    CONFIG_CRYPTO_ZSTD=y
+    CONFIG_ZRAM=y
+    CONFIG_MAGIC_SYSRQ=y
+  '';
+  # hack: drop duplicated entries
+  configfile = pkgs.runCommand "config" { } ''
+    cp "${duo-buildroot-sdk}/build/boards/cv181x/cv1812cp_milkv_duo256m_sd/linux/cvitek_cv1812cp_milkv_duo256m_sd_defconfig" "$out"
+    substituteInPlace "$out" \
+      --replace CONFIG_BLK_DEV_INITRD=y "" \
+      --replace CONFIG_DEBUG_FS=y       "" \
+      --replace CONFIG_VECTOR=y         "" \
+      --replace CONFIG_POWER_RESET=y    "" \
+      --replace CONFIG_RTC_CLASS=y      "" \
+      --replace CONFIG_ZRAM=m           "" \
+      --replace CONFIG_SIGNALFD=n       CONFIG_SIGNALFD=y \
+      --replace CONFIG_TIMERFD=n        CONFIG_TIMERFD=y \
+      --replace CONFIG_EPOLL=n          CONFIG_EPOLL=y
+    cat ${extraconfig} >> "$out"
+  '';
+  kernel = (pkgs.linuxManualConfig {
+    inherit version src configfile;
+    allowImportFromDerivation = true;
+  }).overrideAttrs {
+    preConfigure = ''
+      substituteInPlace arch/riscv/Makefile \
+        --replace '-mno-ldd' "" \
+        --replace 'KBUILD_CFLAGS += -march=$(riscv-march-cflags-y)' \
+                  'KBUILD_CFLAGS += -march=$(riscv-march-cflags-y)_zicsr_zifencei' \
+        --replace 'KBUILD_AFLAGS += -march=$(riscv-march-aflags-y)' \
+                  'KBUILD_AFLAGS += -march=$(riscv-march-aflags-y)_zicsr_zifencei'
+    '';
+  };
+in
+{
+
+  disabledModules = [
+    "profiles/all-hardware.nix"
+  ];
+
+  imports = [
+    "${modulesPath}/installer/sd-card/sd-image.nix"
+  ];
+
+  nixpkgs = {
+    localSystem.config = "x86_64-unknown-linux-gnu";
+    crossSystem.config = "riscv64-unknown-linux-gnu";
+  };
+
+  boot.kernelPackages = pkgs.linuxPackagesFor kernel;
+
+  # neither boot.kernelParams nor boot.consoleLogLevel have any effect here
+  # due to the way the duo images work
+  boot.kernelParams = [ "console=ttyS0,115200" "earlycon=sbi" "riscv.fwsz=0x80000" ];
+  boot.consoleLogLevel = 9;
+
+  boot.initrd.includeDefaultModules = false;
+  boot.initrd.systemd = {
+    # enable = true;
+    # enableTpm2 = false;
+  };
+
+  boot.loader = {
+    grub.enable = false;
+  };
+
+  boot.kernel.sysctl = {
+    "vm.watermark_boost_factor" = 0;
+    "vm.watermark_scale_factor" = 125;
+    "vm.page-cluster" = 0;
+    "vm.swappiness" = 180;
+    "kernel.pid_max" = 4096 * 8; # PAGE_SIZE * 8
+  };
+
+  system.build.dtb = pkgs.runCommand "duo256m.dtb" { nativeBuildInputs = [ pkgs.dtc ]; } ''
+    dtc -I dts -O dtb -o "$out" ${pkgs.writeText "duo256m.dts" ''
+      /include/ "${./prebuilt/cv1812cp_milkv_duo256m_sd.dts}"
+      / {
+        chosen {
+          bootargs = "init=${config.system.build.toplevel}/init ${toString config.boot.kernelParams}";
```

Comment:
If this works then it should automatically boot through stage2?

Reply:
I think it's due to the bootargs set in u-boot taking precedence, I ran into the same problem on my last attempt.

Reply:
Oh! And a correction, I did manage to make the `kernelParams` work, but it requires a fork of the vendor u-boot: https://github.com/NickCao/duo-buildroot-sdk/commit/ccc37b279d32cdf3d5d8b6bf4c0064a630e08f4e

Reply:
Yeah, I guess it doesn't

Reply:
Do you think we should use the buildroot fork?

Reply:
> Do you think we should use the buildroot fork?

How does it differ from the vendor one? I'd prefer one following the distroboot boot flow instead of the current vendor specifc one.

Reply:
Just need some minor modifications for cv181x.

Reply:
> Do you think we should use the buildroot fork?

It's only for the u-boot (fip.bin) part. We can still track the vendor repo for the kernel.

Reply:
Yep, it works.... I generated the prebuilt files from a fork of duo-buildroot-sdk at https://github.com/mcdonc/duo-buildroot-sdk/tree/nixos-riscv